### DISCUSSION TITLE
Unit tests failed after seemlingly unrelated fix for Atracsys wrapper

### DISCUSSION THREAD

**Discussion by atracsys-sbt** [PlusToolkit/PlusLib#822] [Q&A]
Hi,
I'm Sylvain, employee at Atracsys LLC. As mentioned in our first communication, we have been working on fixing compatibility issues between our latest public SDK (4.5.2) and PlusLib. I'm happy to report that I have successfully compiled and tested the current version of PlusLib with various markers on both our FustionTrack (ftk) and SpryTrack (stk) devices.
However, when running PlusLib's BuildAndTest.bat, all unit tests either failed or did not run. I don't think those failures originate from my tweak to the code, however for the sake of being safe before pushing our fix, I would appreciate some help in understanding what's wrong.
Please find attached the files we modified to compare with their current version.
I don't know if it's worth mentionning, but in CMake I unchecked PLUSBUILD_INSTALL_VTK since I already have VTK9.0.1 installed on my machine and leaving it checked triggered a configuration error (Qt5 could not be found despite a correct QT5_DIR).
Regards,
S.
[plus_fix_4.5.2.zip](https://github.com/PlusToolkit/PlusLib/files/6663621/plus_fix_4.5.2.zip)

**Comment by lassoan** [comment]
Did the test pass before the change?
Do you see errors in the event log (something about dll not found)?

Instead of attaching a zip file, please create a pull request with the proposed changes. Thank you!

**Comment by atracsys-sbt** [comment]
So, I created a fresh clone of PlusLib, configured it with PLUSBUILD_INSTALL_VTK **unchecked**, compiled it and ran PlusLib-bin/BuildAndTest.bat. All tests either failed or were skipped, as shown in the attached log file 1.
Then, I reconfigured PlusLib with PLUSBUILD_INSTALL_VTK **checked**, compiled it and ran the same script again. The tests showed mostly segmentation faults (see attached log file 2).
Keep in mind that so far I haven't touched the code, so I think my problems arise from a conflict between the VTK already installed on my computer and the one provided by PlusLib, though seemingly from the same version (9.0.1).

[1 BuildTest PlusLib.txt](https://github.com/PlusToolkit/PlusLib/files/6669241/1.BuildTest.PlusLib.txt)
[2 BuildTest PlusLib vtkInstalled.txt](https://github.com/PlusToolkit/PlusLib/files/6669242/2.BuildTest.PlusLib.vtkInstalled.txt)

 I created the pull request with the fix we propose for the Atracsys wrapper.
 Thank your for your help

**Reply by lassoan** [reply]
> VTK already installed on my computer

VTK should never be "installed" on a Windows computer (put in a folder that is in the path or in a system folder), because you can compile VTK many different ways, so if an application loads these other VTK DLLs then it will crash. The problem is particularly severe if you install it in some system location because then it will be found even if the correct VTK folder is added to the path (this could break VTK in virtual Python environments, etc). You would need to use side-by-side assemblies and manifests to isolate different VTK versions, but I don't think these are supported by VTK's build system.

I would strongly recommend to remove all VTK DLLs from shared locations to prevent breaking various VTK-based applications and libraries.

**Reply by atracsys-sbt** [reply]
I understand. After removing the VTK installation from C:\Program Files as well as from the system path variable, I tried again the unit tests and again all but one failed (in the comparison between output and ground truth), as you can see in the attached log.
[3 BuildTest PlusLib vtk removed from path.txt](https://github.com/PlusToolkit/PlusLib/files/6670757/3.BuildTest.PlusLib.vtk.removed.from.path.txt)

**Reply by lassoan** [reply]
Please check in the Windows Event Viewer which DLL is not found. It may give a useful hint.

@Sunderlandkyl I have not run Plus tests for a while. Is BuildAndTest.bat still supported or do you usually run the tests in a different way?

**Reply by Sunderlandkyl** [reply]
BuildAndTest is still supported. The nightly builds all still use PlusBuildAndTest/BuildAndTest.

**Reply by atracsys-sbt** [reply]
I ran the test again (with the same result) and I can't find anything in the Event Viewer that occurred during the test. Could you be more specific as to what log would show the entry of a missing dll ?

**Reply by atracsys-sbt** [reply]
> 
> 
> BuildAndTest is still supported. The nightly builds all still use PlusBuildAndTest/BuildAndTest.

I just want to point out that running PlusBuild/BuildAndTest.bat was ok, PlusApp-bin/BuildAndTest.bat was ok, only PlusLib-bin/BuildAndTest.bat fails.

**Reply by lassoan** [reply]
You can add `-VV` to the [line that launches CTest](https://github.com/PlusToolkit/PlusLib/blob/27163f638d133e240ae2967719d93c1fd02648a6/BuildAndTest.bat.in#L45). It will make it print the complete command of each test. If you run that command in a terminal then you may get more specific information in an error popup and more information may be logged into the Event Viewer (in Application or System section).

**Reply by atracsys-sbt** [reply]
Ok, so my first mistake was running the tests in PowerShell, because once I tried in a regular windows terminal, I immediately got a popup error indicating what dll were missing.
Manually running test 2 (EditSequenceFileTrim), the popup informed me that I was missing:
vtkIOXMLParser-9.0.dll, vtkCommonDataModel-9.0.dll, vtkCommonTransforms-9.0.dll and vtkCommonMath-9.0.dll
which are indeed not along the other vtk dlls in bin/Release... Which also explains why I had different outputs (segfaults) when vtk was installed elsewhere.

**Reply by lassoan** [reply]
Are these DLLs in the VTK build tree, just not copied to the bin folder?
As far as I remember, the copying happens automatically, but maybe something broke due to the large build system rework in VTK9. Maybe the issue is that you changed VTK and you did not restart a completely new build from scratch.

**Reply by atracsys-sbt** [reply]
So I restarted from fresh again, this time leaving the INSTALL_VTK checked and I ran into the same issue that prevented me to compile the first time around. This is the first of three similar errors from the compilation output:
```
12>CMake Error at D:/dev-ext/build/PlusBuild_clean/vtk-int/lib/cmake/vtk-9.0/VTK-vtk-module-find-packages.cmake:115 (find_package):
12>  By not providing "FindQt5.cmake" in CMAKE_MODULE_PATH this project has
12>  asked CMake to find a package configuration file provided by "Qt5", but
12>  CMake did not find one.
12>
12>  Could not find a package configuration file provided by "Qt5" (requested
12>  version 5.15) with any of the following names:
12>
12>    Qt5Config.cmake
12>    qt5-config.cmake
12>
12>  Add the installation prefix of "Qt5" to CMAKE_PREFIX_PATH or set "Qt5_DIR"
12>  to a directory containing one of the above files.  If "Qt5" provides a
12>  separate development package or SDK, be sure it has been installed.
12>Call Stack (most recent call first):
12>  D:/dev-ext/build/PlusBuild_clean/vtk-int/lib/cmake/vtk-9.0/vtk-config.cmake:131 (include)
12>  CMakeLists.txt:47 (find_package)
```
What's weird is that I did provide a valid Qt5_DIR and there is indeed a Qt5Config.cmake.
From what I gather, I could compile PlusBuild as long as I left INSTALL_VTK unchecked, but at the cost of not having all dll's copied, hence the problem with unit tests.

**Reply by lassoan** [reply]
`D:/dev-ext/build/PlusBuild_clean` path may be too long. Please follow the [build instructions](https://github.com/PlusToolkit/PlusBuild/blob/master/Docs/BuildInstructionsWindows.md#build-process) exactly. If you ignore any instruction at any point then you may run into errors later and it is not guaranteed that the error message will be intuitive.

From the error message it seems that you don't use PlusBuild to build Plus and all its dependencies (if you used PlusBuild then Qt not found error would have been reported as `This project requires Qt5 for building PlusApp but Qt was not found. Please specify Qt5_DIR or turn off PLUSBUILD_BUILD_PLUSAPP/PLUSBUILD_BUILD_PLUSLIB_WIDGETS options.`). For your first build, use PlusBuild. Later, if you are very confident then you can start building all the dependencies on your own.

What Qt version did you download? Did you download 32-bit or 64-bit Qt binaries? Do you build a 32-bit or 64-bit Plus?

**Reply by atracsys-sbt** [reply]
Thank you for your reply once again.
So, I did follow carefully the build instructions (notably a much shorter path) and I still got the exact same error during the compilation. Also, I've always cloned and compiled from PlusBuild, not PlusLib.
I have Qt5.15 x64 installed and I'm trying to compile Plus in x64 alike.

**Reply by lassoan** [reply]
Did you choose VTK version 9 when you configured the project? Did yiu change any other build option in CMake? Which Visual Studio version and toolset version do you use? Please attach a screenshot of your Qt installer (where you configure the installed Qt components) so that we can compare it with working configurations.

**Reply by atracsys-sbt** [reply]
The first time I configure it, after choosing Visual Studio 16 2019 as my solution generator, there is an error because of Qt5_DIR being unspecified, so I enter the path (C:/Qt/5.15.2/msvc2019_64/lib/cmake/Qt5). Then I configure again and get another error relative to vtk requiring version 9.0 at least because of Qt5.15, so I select 9.0.1 in EXTERNAL_VTK_VERSION. Then I configure and generate.
Please find attached a screenshot of all Qt components I have installed.
![qt5_install](https://user-images.githubusercontent.com/85866864/122761052-1a207d00-d29c-11eb-9cd9-6defb9636111.PNG)

**Reply by lassoan** [reply]
I could reproduce the build error using VTK9 and Qt5.

@Sunderlandkyl does the dashboard use VTK9? Have you tried to build Plus with VTK9?

I've submitted two poull requests ([1](https://github.com/PlusToolkit/PlusBuild/pull/73) and [2](https://github.com/IGSIO/IGSIO/pull/24)), which fixes the error for me.

@Sunderlandkyl please have a look at them. Especially the IGSIO is very simplistic, as it just hardcodes setting Qt directory, regardless of VTK version.

It would be also better to understand this and avoid passing Qt manually everywhere.

**Reply by Sunderlandkyl** [reply]
Dashboard still uses VTK 8. I haven't tried building Plus with VTK 9.

It looks like Qt5_DIR was redefined in VTK8 in a number of generated CMake files.

Ex. in VTK8 vtkGUISupportQt.cmake:
```
if(NOT Qt5_DIR)
  set(Qt5_DIR "C:/Qt/Qt5.10.1/5.10.1/msvc2015_64/lib/cmake/Qt5")
endif()
find_package(Qt5 REQUIRED QUIET COMPONENTS Widgets)
```

Those calls aren't present in VTK9, so that seems to be the culprit.

**Reply by lassoan** [reply]
The fixes have been merged, so PlusLib build with VTK9 should now work well.

**Reply by atracsys-sbt** [reply]
I can confirm that I can successfully compile PlusLib with VTK9 now. That said, despite having INSTALL_VTK checked, there are still missing vtk dll's in /bin and thus the unit tests still fail.

**Reply by lassoan** [reply]
Thanks for testing. Plus official builds use VTK8. @adamrankin has implemented some VTK9 support in Plus, but apparently it is incomplete. If you want to use VTK9 then you need to copy the VTK dlls to the PlusBuild-bin\bin\Release folder manually.

@adamrankin could you fix copying/installation of VTK dlls into PlusBuild-bin\bin\Release folder?
